File: ios_app/CatchMeNearby/Features/UserProfile/ViewModels/ProfileViewModel.swift
import SwiftUI
import FirebaseFirestore
import FirebaseFirestoreSwift // For Codable support with Firestore
import FirebaseStorage
import FirebaseAuth // To get current user UID

class ProfileViewModel: ObservableObject {
    @Published var userProfile: UserProfile?
    // Fields bound to UI, initialized from userProfile
    @Published var username: String = ""
    @Published var bio: String = ""
    @Published var locationSharingEnabled: Bool = true
    @Published var photoItems: [String] = [] // Stores URLs of photos or local identifiers for upload

    @Published var isLoading: Bool = false
    @Published var errorMessage: String?
    @Published var saveSuccess: Bool = false


    private var db = Firestore.firestore()
    // private var storage = Storage.storage() // Uncomment if actual uploads are implemented
    private var currentUserID: String? { Auth.auth().currentUser?.uid }

    private var cancellables = Set<AnyCancellable>() // Corrected var name
    private var locationService: LocationService? // To be injected or received from environment

    // init() called by @StateObject or @ObservedObject
    // If locationService is passed via environment, it won't be available in init directly
    // unless this ViewModel is also an EnvironmentObject or receives it as a parameter.
    // For now, let's assume it will be set up after init.
    // A better pattern might be to pass LocationService in init.
    // init(locationService: LocationService) {
    //    self.locationService = locationService
    //    // ... existing init logic ...
    // }

    init() {
        // Combine pipeline to react to userProfile changes and update bound fields
        $userProfile
            .receive(on: RunLoop.main)
            .sink { [weak self] profile in
                guard let self = self, let profile = profile else { return }
                self.username = profile.username
                self.bio = profile.bio ?? ""
                self.locationSharingEnabled = profile.locationSharingEnabled
                self.photoItems = profile.photoURLs ?? []
            }
            .store(in: &cancellables)

        // fetchUserProfile() // Call fetchUserProfile when currentUserID is known, e.g. onAppear or when auth state changes.
                           // Or if locationService is passed in init:
                           // self.locationService = locationService
                           // fetchUserProfile()
    }

    // Call this method when the ViewModel appears or when the LocationService becomes available.
    func setup(locationService: LocationService) {
        self.locationService = locationService
        // Now that locationService is available, we can fetch the profile
        // and then update location service with profile settings.
        if userProfile == nil { // Fetch only if not already loaded
             fetchUserProfile()
        } else {
             // If profile is already loaded (e.g. from a cache or previous fetch),
             // still update locationService with current settings.
             self.locationService?.updateUserProfileSettings(self.userProfile)
        }

        // Listen for app lifecycle notifications to manage locationService
        NotificationCenter.default.publisher(for: UIApplication.didBecomeActiveNotification)
            .sink { [weak self] _ in self?.locationService?.appDidEnterForeground() }
            .store(in: &cancellables)

        NotificationCenter.default.publisher(for: UIApplication.didEnterBackgroundNotification)
            .sink { [weak self] _ in self?.locationService?.appDidEnterBackground() }
            .store(in: &cancellables)
    }


    func fetchUserProfile() {
        guard let userID = currentUserID else {
            errorMessage = "User not logged in."
            // Potentially stop location updates if user logs out.
            // locationService?.updateUserProfileSettings(nil) // This would stop it.
            return
        }
        isLoading = true
        db.collection("users").document(userID).getDocument { [weak self] documentSnapshot, error in
            // Defer is not needed here as isLoading is set at start and end of relevant blocks
            if let error = error {
                self?.isLoading = false
                self?.errorMessage = "Error fetching profile: \(error.localizedDescription)"
                return
            }
            guard let document = documentSnapshot, document.exists else {
                self?.isLoading = false
                // This case might mean a new user whose profile wasn't created yet by signup logic.
                // Or, an actual error. For now, treat as an error or missing profile.
                // Consider creating a default profile here if that's the desired flow.
                self?.errorMessage = "Profile document does not exist. Please complete signup or contact support."
                // No profile, so location sharing is effectively off for this non-existent user.
                // Update location service with nil profile to stop any potential updates.
                self?.locationService?.updateUserProfileSettings(nil)
                return
            }
            do {
                let fetchedProfile = try document.data(as: UserProfile.self)
                self?.userProfile = fetchedProfile
                // The sink for $userProfile will update other @Published properties (username, bio, etc.)
                self?.errorMessage = nil
                // After fetching profile, update LocationService with the settings
                self?.locationService?.updateUserProfileSettings(fetchedProfile)
            } catch {
                self?.errorMessage = "Error decoding profile: \(error.localizedDescription)"
                self?.locationService?.updateUserProfileSettings(nil) // Stop on error
            }
            self?.isLoading = false // Ensure isLoading is set to false in all paths
        }
    }

    func saveProfile() {
        guard let userID = currentUserID else {
            errorMessage = "User not logged in or profile not loaded."
            return
        }
        isLoading = true
        saveSuccess = false // Reset save success status

        // Create a dictionary for updates to only send changed values if preferred,
        // or update the local userProfile object and set it entirely.
        // For simplicity with Codable, updating the local userProfile object and re-setting it is easier.

        var profileToUpdate = self.userProfile ?? UserProfile(id: userID, username: self.username) // Create new if nil (shouldn't happen if fetch worked)

        profileToUpdate.username = username
        profileToUpdate.bio = bio.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? nil : bio.trimmingCharacters(in: .whitespacesAndNewlines)
        profileToUpdate.locationSharingEnabled = locationSharingEnabled
        profileToUpdate.photoURLs = photoItems // Assumes photoItems holds the final URLs
        profileToUpdate.updatedAt = Timestamp(date: Date()) // Firestore server timestamp preferred, but client can set for optimistic UI

        do {
            // Using merge: true to only update fields present in the model, good for partial updates
            // or if the model might not have all fields (e.g. createdAt).
            // If UserProfile is complete, merge: false (the default) would overwrite the document.
            // Given we fetch then update, merge: true is safer.
            try db.collection("users").document(userID).setData(from: profileToUpdate, merge: true) { [weak self] error in
                guard let self = self else { return }
                self.isLoading = false
                if let error = error {
                    self.errorMessage = "Error saving profile: \(error.localizedDescription)"
                } else {
                    self.errorMessage = nil
                    self.saveSuccess = true
                    self.userProfile = profileToUpdate // Update local cache
                    print("Profile saved successfully.")
                    // After saving, especially if locationSharingEnabled changed, update LocationService
                    self.locationService?.updateUserProfileSettings(profileToUpdate)
                }
            }
        } catch {
            self.isLoading = false
            self.errorMessage = "Error encoding profile for save: \(error.localizedDescription)"
        }
    }

    // Expose a way to request location permission if needed from UI,
    // though LocationService tries to handle it internally.
    func requestLocationPermissionIfNeeded() {
        locationService?.requestPermission()
    }

    func addPhoto(placeholderURL: String) {
        if photoItems.count < 7 {
            photoItems.append(placeholderURL)
            // In a real app: trigger image picker, upload, get URL, then add.
        } else {
            errorMessage = "Maximum of 7 photos allowed."
        }
    }

    func removePhoto(at offsets: IndexSet) {
        photoItems.remove(atOffsets: offsets)
        // In a real app: if URLs point to Firebase Storage, also delete the file from Storage.
    }
}

File: android_app/app/src/main/java/com/catchmenearby/features/profile/viewmodel/ProfileViewModel.java
package com.catchmenearby.features.profile.viewmodel;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.FirebaseFirestore;
// import com.google.firebase.storage.FirebaseStorage; // For photo uploads later
// import com.google.firebase.storage.StorageReference; // For photo uploads later
import com.catchmenearby.features.profile.model.UserProfile;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

public class ProfileViewModel extends ViewModel {
    private FirebaseFirestore db = FirebaseFirestore.getInstance();
    // private FirebaseStorage storage = FirebaseStorage.getInstance(); // For photo uploads
    private FirebaseAuth auth = FirebaseAuth.getInstance();

    private MutableLiveData<UserProfile> userProfileLiveData = new MutableLiveData<>();
    // MutableLiveData for individual editable fields, two-way bound or updated from UI
    private MutableLiveData<String> username = new MutableLiveData<>();
    private MutableLiveData<String> bio = new MutableLiveData<>();
    private MutableLiveData<Boolean> locationSharingEnabled = new MutableLiveData<>(true); // Default
    private MutableLiveData<List<String>> photoItems = new MutableLiveData<>(new ArrayList<>());

    private MutableLiveData<Boolean> isLoading = new MutableLiveData<>(false);
    private MutableLiveData<String> errorMessage = new MutableLiveData<>();
    private MutableLiveData<Boolean> saveSuccess = new MutableLiveData<>(false);

    public ProfileViewModel() {
        fetchUserProfile();
    }

    // LiveData getters for observing in Activity/Fragment
    public LiveData<UserProfile> getUserProfileLiveData() { return userProfileLiveData; }
    public LiveData<String> getUsernameState() { return username; } // Changed name to avoid conflict
    public LiveData<String> getBioState() { return bio; } // Changed name to avoid conflict
    public LiveData<Boolean> getLocationSharingEnabledState() { return locationSharingEnabled; } // Changed name
    public LiveData<List<String>> getPhotoItemsState() { return photoItems; } // Changed name
    public LiveData<Boolean> getIsLoading() { return isLoading; }
    public LiveData<String> getErrorMessage() { return errorMessage; }
    public LiveData<Boolean> getSaveSuccess() { return saveSuccess; }

    // Methods to update individual fields from UI (e.g., after EditText changes)
    public void updateUsername(String uname) { username.setValue(uname); }
    public void updateBio(String userBio) { bio.setValue(userBio); }
    public void updateLocationSharingEnabled(boolean enabled) { locationSharingEnabled.setValue(enabled); }


    public void fetchUserProfile() {
        FirebaseUser currentUser = auth.getCurrentUser();
        if (currentUser == null) {
            errorMessage.setValue("User not logged in.");
            return;
        }
        isLoading.setValue(true);
        db.collection("users").document(currentUser.getUid()).get()
            .addOnSuccessListener(documentSnapshot -> {
                isLoading.setValue(false);
                if (documentSnapshot.exists()) {
                    UserProfile profile = documentSnapshot.toObject(UserProfile.class);
                    userProfileLiveData.setValue(profile); // This is the "master" copy
                    if (profile != null) {
                        // Initialize editable LiveData from the fetched profile
                        username.setValue(profile.getUsername());
                        bio.setValue(profile.getBio() == null ? "" : profile.getBio());
                        locationSharingEnabled.setValue(profile.isLocationSharingEnabled());
                        photoItems.setValue(profile.getPhotoURLs() != null ? new ArrayList<>(profile.getPhotoURLs()) : new ArrayList<>());
                        errorMessage.setValue(null);
                    } else {
                         errorMessage.setValue("Failed to parse profile data.");
                    }
                } else {
                    errorMessage.setValue("Profile document does not exist. Please complete signup or contact support.");
                }
            })
            .addOnFailureListener(e -> {
                isLoading.setValue(false);
                errorMessage.setValue("Error fetching profile: " + e.getMessage());
            });
    }

    public void saveProfile() {
        FirebaseUser currentUser = auth.getCurrentUser();
        // Use a local copy of userProfile or construct a new one for saving
        // This helps if userProfileLiveData.getValue() is null for some reason
        UserProfile profileToSave;
        if (userProfileLiveData.getValue() != null) {
            profileToSave = userProfileLiveData.getValue();
        } else {
            // Fallback or error, ideally profile should be loaded
            profileToSave = new UserProfile();
            profileToSave.setId(currentUser != null ? currentUser.getUid() : "error_no_uid");
        }


        if (currentUser == null) {
            errorMessage.setValue("User not logged in. Cannot save profile.");
            return;
        }
        isLoading.setValue(true);
        saveSuccess.setValue(false); // Reset save success status

        DocumentReference profileRef = db.collection("users").document(currentUser.getUid());

        // Update the profile object with values from the UI-bound LiveData
        profileToSave.setUsername(username.getValue());
        String bioValue = bio.getValue();
        profileToSave.setBio(bioValue != null && bioValue.trim().isEmpty() ? null : bioValue);
        profileToSave.setLocationSharingEnabled(locationSharingEnabled.getValue() != null ? locationSharingEnabled.getValue() : true);
        profileToSave.setPhotoURLs(photoItems.getValue()); // Assumes photoItems LiveData holds the current list of URLs
        // profileToSave.setUpdatedAt(new Date()); // Let @ServerTimestamp handle this

        profileRef.set(profileToSave) // Overwrites the document with the content of profileToSave
                                      // Use .update() for specific fields if that's preferred.
                                      // If UserProfile.java has @ServerTimestamp on updatedAt, it will be handled by server.
            .addOnSuccessListener(aVoid -> {
                isLoading.setValue(false);
                errorMessage.setValue(null); // Clear any previous error
                saveSuccess.setValue(true); // Signal success
                userProfileLiveData.setValue(profileToSave); // Update the "master" LiveData
                System.out.println("Profile saved successfully for UID: " + currentUser.getUid());
            })
            .addOnFailureListener(e -> {
                isLoading.setValue(false);
                errorMessage.setValue("Error saving profile: " + e.getMessage());
                saveSuccess.setValue(false);
            });
    }

    public void addPhotoPlaceholder(String placeholderURL) {
        List<String> currentPhotos = photoItems.getValue();
        if (currentPhotos == null) {
            currentPhotos = new ArrayList<>();
        }
        if (currentPhotos.size() < 7) {
            ArrayList<String> updatedPhotos = new ArrayList<>(currentPhotos);
            updatedPhotos.add(placeholderURL);
            photoItems.setValue(updatedPhotos);
        } else {
            errorMessage.setValue("Maximum of 7 photos allowed.");
        }
    }

    public void removePhotoAtIndex(int index) {
         List<String> currentPhotos = photoItems.getValue();
         if (currentPhotos != null && index >= 0 && index < currentPhotos.size()) {
             ArrayList<String> updatedPhotos = new ArrayList<>(currentPhotos);
             updatedPhotos.remove(index);
             photoItems.setValue(updatedPhotos);
         }
    }

    // Method to get current user ID, useful for ViewProfileActivity logic
    public String getCurrentUserId() {
        FirebaseUser currentUser = auth.getCurrentUser();
        return (currentUser != null) ? currentUser.getUid() : null;
    }

    // Method to fetch a specific user's profile by their ID
    public void fetchSpecificUserProfile(String userId) {
        if (userId == null || userId.isEmpty()) {
            errorMessage.setValue("User ID is invalid.");
            return;
        }
        isLoading.setValue(true);
        // Clear existing profile data before fetching new one, or indicate it's for a different user
        // userProfileLiveData.setValue(null); // Optional: clear previous profile

        db.collection("users").document(userId).get()
            .addOnSuccessListener(documentSnapshot -> {
                isLoading.setValue(false);
                if (documentSnapshot.exists()) {
                    UserProfile profile = documentSnapshot.toObject(UserProfile.class);
                    userProfileLiveData.setValue(profile); // This will be observed by ViewProfileActivity
                    if (profile != null) {
                        // These are for editing, ViewProfileActivity will use userProfileLiveData directly
                        // username.setValue(profile.getUsername());
                        // bio.setValue(profile.getBio() == null ? "" : profile.getBio());
                        // locationSharingEnabled.setValue(profile.isLocationSharingEnabled());
                        // photoItems.setValue(profile.getPhotoURLs() != null ? new ArrayList<>(profile.getPhotoURLs()) : new ArrayList<>());
                        errorMessage.setValue(null);
                    } else {
                        errorMessage.setValue("Failed to parse profile data for user: " + userId);
                    }
                } else {
                    errorMessage.setValue("Profile document does not exist for user: " + userId);
                }
            })
            .addOnFailureListener(e -> {
                isLoading.setValue(false);
                errorMessage.setValue("Error fetching profile for user " + userId + ": " + e.getMessage());
            });
    }
}
